packages/core/database/src/utils/remove-orphan-morph-types.ts

```diff
@@ -0,0 +1,70 @@
+import type { Database } from '..';
+
+/**
+ * Removes morph relation data with invalid or non-existent morph types.
+ *
+ * This function iterates over the database metadata to identify morph relationships
+ * (relations with a `joinTable` containing the specified pivot column) and removes
+ * any entries in the relation's join table where the morph type is invalid:
+ * - Morph types without corresponding metadata.
+ * - Morph types pointing to tables that no longer exist in the database.
+ *
+ * Note: This function does not check for orphaned IDs, only orphaned morph types.
+ *
+ * @param db - The database object containing metadata and a Knex connection.
+ * @param pivot - The name of the column in the join table representing the morph type.
+ */
+export const removeOrphanMorphTypes = async (db: Database, pivot: string) => {
+  const allTables = await db.dialect.schemaInspector.getTables();
+
+  for (const model of db.metadata.values()) {
+    const attributes = Object.values(model.attributes || {}).filter(
+      (attribute) =>
+        attribute.type === 'relation' &&
+        'joinTable' in attribute &&
+        'name' in attribute.joinTable &&
+        'target' in attribute &&
+        'pivotColumns' in attribute.joinTable &&
+        attribute.joinTable.pivotColumns.includes(pivot)
+    );
+
+    for (const attribute of attributes) {
+      if (!('joinTable' in attribute)) {
```

Comment:
I supposed it's just to fix the typings ?

Reply:
yeah, it narrows down the attribute type

Reply:
yeah it should already be narrowed down above. I recommend you just loop over the attributes once and do the filter in the for loop cleanly instead

Reply:
I think it's a typescript parser issue, it wasn't recognizing it from above. But I can keep playing with it and probably get that removed

Reply:
You shoul use a type assertion for this 👍  helps the typescript parsing & also makes the code quite clean